### PR TITLE
detach returned tensors from source tensors in tensor constructors

### PR DESCRIPTION
**Pull request by weiyangfb**
- fix PR https://github.com/pytorch/pytorch/pull/11061 by moving `detach_()` and `set_requires_grad()` to `torch.tensor_ctor()` and `tensor.new_tensor`, and also removed warnings and `args_requires_grad` from `internal_new_from_data `
- with this patch, the returned tensor from `tensor_ctor()` and `new_tensor` will be detached from source tensor, and set requires_grad based on the input args
- `torch.as_tensor` retains its behavior as documented

@gchanan @apaszke

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -506,18 +499,25 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<4> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
+    PyObject* data = r.pyobject(0);
+    if (THPVariable_Check(data)) {
+      PyErr_WarnEx(PyExc_UserWarning,
```
Comment (SsnL) [inline torch/csrc/utils/tensor_new.cpp]: I agree with @gchanan 's comment on last PR that this should mention `as_tensor`.

File: test/test_torch.py [hunk 2]
```diff
@@ -2397,29 +2397,23 @@ def test_tensor_factory(self):
             self.assertEqual(5., res1[0].item())
 
     def test_tensor_factory_copy_var(self):
```
Comment (SsnL) [inline test/test_torch.py]: Do the changes here test the bug introduced in #11061 ? I don't see any `as_tensor` calls.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: I see, thanks a lot for catching this! So how should I do this correctly, can I:
1. raise similar warnings for `as_tensor` when `type_inference` is false (copy construct)
2. no need to detach the new tensor

File: test/test_torch.py (hunk 2 above)
Comment (weiyangfb) [reply test/test_torch.py]: No it doesn't, yeah I should add tests for `as_tensor` as well

File: test/test_torch.py [hunk 3]
```diff
@@ -2397,29 +2397,28 @@ def test_tensor_factory(self):
...
+        # test tensor.new_tensor()
+        copy = torch.randn(1)
+        check_copy(copy.new_tensor(source), True, False)
+        check_copy(copy.new_tensor(source, requires_grad=False), True, False)
+        check_copy(copy.new_tensor(source, requires_grad=True), True, True)
+
+        # test torch.as_tensor()
+        source = source.add(1)  # make source non-leaf
+        check_copy(torch.as_tensor(source), source.is_leaf, source.requires_grad)
+        check_copy(torch.as_tensor(source, dtype=torch.float), True, False)
```
Comment (gchanan) [inline test/test_torch.py]: what is this testing?  The type conversion case?  It would be more clear if you explicitly passed a different dtype to `torch.randn` so we don't have to know what the default dtype is.

File: torch/csrc/utils/tensor_new.cpp [hunk 4]
```diff
@@ -546,11 +576,21 @@ Tensor new_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
...
+    }
+
+    bool args_requires_grad = r.toBool(3);
+    auto new_tensor = new_from_data_copy(
                typeWithDefault(r, 1, 2, type),
                r.deviceOptional(2),
-               r.pyobject(0))
-        .set_requires_grad(r.toBool(3));
+               data);
+    new_tensor.detach_(); // making new_tensor a leaf node
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: nit: I would say 'ensure new_tensor is a leaf node' (it may already be).

File: torch/csrc/utils/tensor_new.cpp [hunk 5]
```diff
@@ -546,11 +576,21 @@ Tensor new_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<4> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
-    return new_from_data_copy(
+    PyObject* data = r.pyobject(0);
+    if (THPVariable_Check(data)) {
+      PyErr_WarnEx(PyExc_UserWarning,
+        "To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() "
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: It seems okay to put the warning in internal_new_from_data (you can pass in the name of the function).  The problem with the previous implementation was that args_requires_grad wasn't respected across all code paths, but the name of the function can easily be respected across all code paths.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (gchanan) [reply torch/csrc/utils/tensor_new.cpp]: This is a good question.  My opinion is that we don't need to go through the whole warning song and dance with `as_tensor` because it doesn't have a `requires_grad` argument (precisely for the reason mentioned).  We just need to add to the `as_tensor` document that it either returns the tensor passed in or it returns a new Tensor with `requires_grad = False` (it follows that you don't need to call detach and detach isn't preferred from the fact that `requires_grad = False`).

File: torch/csrc/utils/tensor_new.cpp [hunk 6]
```diff
@@ -506,18 +499,25 @@ Tensor tensor_ctor(const Type& type, PyObject* args, PyObject* kwargs) {
...
                r.deviceOptional(2),
-               r.pyobject(0),
+               data,
                true,
                true,
-               type_inference,
-               args_requires_grad)
-        .set_requires_grad(r.toBool(3));
+               type_inference);
+    new_tensor.detach_(); // making new_tensor a leaf node
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: prefer 'ensure' to 'making' -- new_tensor may already be a leaf node.

File: torch/csrc/utils/tensor_new.cpp [hunk 7]
```diff
@@ -531,9 +531,39 @@ Tensor as_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<3> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
+    at::optional<Device> device_opt = r.deviceOptional(2);
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: as mentioned above, I don't think you need any of this.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: Sounds great! Thanks for the clarifications! Indeed `as_tensor` doesn't have a `requires_grad` in args and so no need any detach or warning in the code path. I will add a note to the docs accordingly.

File: test/test_torch.py (hunk 3 above)
Comment (weiyangfb) [reply test/test_torch.py]: oh, right, I will make the dtype explicit.

File: torch/csrc/utils/tensor_new.cpp (hunk 7 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: right, I will remove them all together

File: test/test_torch.py [hunk 8]
```diff
@@ -2397,29 +2397,28 @@ def test_tensor_factory(self):
...
-        source = torch.randn(5, 5, requires_grad=True)
-        copy = torch.tensor(source, requires_grad=False)
-        self.assertEqual(copy.data, source.data)
-        self.assertTrue(source.requires_grad)
-        self.assertTrue(copy.is_leaf)
-        self.assertFalse(copy.requires_grad)
+        def check_copy(copy, copy_is_leaf, copy_requires_grad):
+            self.assertEqual(copy.data, source.data)
+            self.assertTrue(copy.is_leaf == copy_is_leaf)
+            self.assertTrue(copy.requires_grad == copy_requires_grad)
```
Comment (apaszke) [inline test/test_torch.py]: Can we also check that `.data_ptr`s are equal/different (depending on the case)?

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -449,6 +449,14 @@ def parse_kwargs(desc):
 will be performed.  Similarly, if the data is an ``ndarray`` of the corresponding `dtype` and the `device` is the cpu,
 no copy will be performed.
 
+.. warning::
+
+    When attr:`data` is a tensor `x` with different attr:`dtype` or attr:`device`,
+    :func:`as_tensor()` reads out 'the data' from whatever it is passed,
+    and constructs a leaf variable. Therefore ``tensor.as_tensor(x, dtype=dtype, device=device)``
+    is equivalent to ``x.clone().detach().to(dtype, device)``.
+    The equivalents using ``clone()`` and ``detach()`` are recommended.
```
Comment (apaszke) [inline torch/_torch_docs.py]: Hmmm I'm not sure if that's a good decision. I'd expect `torch.as_tensor(tensor)` to return `tensor` unmodified (and retaining its autograd history).

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/utils/tensor_new.cpp]: I'd say this is confusing. Shouldn't `torch.as_tensor(tensor)` be differentiable?

File: torch/_torch_docs.py (hunk 9 above)
Comment (apaszke) [reply torch/_torch_docs.py]: I also don't think it matches what @gchanan said above. There's no `clone()` happening in `as_tensor`.

File: torch/csrc/utils/tensor_new.cpp [hunk 10]
```diff
@@ -531,9 +531,39 @@ Tensor as_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
...
+    const Type & type_default = typeWithDefault(r, 1, 2, type);
+    PyObject* data = r.pyobject(0);
     bool type_inference = r.isNone(1);
-    return internal_new_from_data(
-        typeWithDefault(r, 1, 2, type), r.deviceOptional(2), r.pyobject(0), false, false, type_inference);
+    bool is_copy = false;
+
+    if (THPVariable_Check(data)) {
```
Comment (apaszke) [inline torch/csrc/utils/tensor_new.cpp]: Why did this code get so much more complicated? The whole reason `internal_from_data` exists is to avoid repeating those things

**Review by apaszke — CHANGES_REQUESTED** [review]
Semantics of `torch.as_tensor` need to be clarified.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (fmassa) [reply torch/csrc/utils/tensor_new.cpp]: Yes, I think it should be differentiable.

Comment (SsnL) [reply torch/csrc/utils/tensor_new.cpp]: Maybe. I'd argue that if users do `as_tensor(something)`, they usually don't know (or didn't bother to figure out) if `something` is a tensor or not, in which case they certainly wouldn't be thinking to themselves "hmm the gradients will glowing back through here, and then ....".

File: test/test_torch.py (hunk 8 above)
Comment (weiyangfb) [reply test/test_torch.py]: Thanks! I will add that

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: @apaszke @fmassa yes, for sure `torch.as_tensor(tensor)` is differentiable. I was wonder if `torch.as_tensor(tensor, dtype=dtype, device=device)`, where dtype and device are different to tensor's, should be differentiable?

File: torch/csrc/utils/tensor_new.cpp (hunk 10 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: yeah, I put this here mainly was because of the confusion on whether or not `torch.as_tensor(tensor, dtype=dtype, device=device)` should be detached. If yes, than I will still need a way to figure out whether or not a new tensor is copy constructed.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (gchanan) [reply torch/csrc/utils/tensor_new.cpp]: ` yes, for sure torch.as_tensor(tensor) is differentiable.`

Well, that's clearly true because it's the same tensor.

The behavior of `torch.as_tensor(tensor, dtype=dtype, device=device` is a good question.  I guess my thought process is you'd call `as_tensor` on heterogeneous inputs (i.e. some things are tensors, some are lists, etc).  So as @SsnL says, you probably don't care about gradients flowing back through.  If you did, you'd call `to`.  It also seems a little strange that `torch.tensor` isn't differentiable, but `torch.as_tensor` is.  I bet I could be convinced of the opposite though! @apaszke ? @fmassa ?

File: torch/_torch_docs.py (hunk 9 above)
Comment (weiyangfb) [reply torch/_torch_docs.py]: Yes, agreed upon the point on `torch.as_tensor(tensor)`! Sorry I haven't push the update to address @gchanan comments. Also, for `torch.as_tensor(tensor, dtype=dtype, device=device)`, what should the copy constructed tensor look like? Should it be made non-differentiable?

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: @fmassa Commented on the workchat, I am reposting it here to drive the discussion:

one way of writing code that works for both tensors and numbers / numpy arrays is to always put a as_tensor right in the beginning of the function. So I could see this as a very common pattern:
```
def exp(x):
    x = torch.as_tensor(x)
    return x.exp()
```
I btw think that this is something that we should consider doing in our functions, so that `torch.exp(1.0)` works.

Thus, I would argue that `torch.as_tensor(x)` should not break the computation graph if x is a tensor that requires grad

Comment (fmassa) [reply torch/csrc/utils/tensor_new.cpp]: Thanks for reposting my message @weiyangfb !
So, to expand a bit more on the previous message, if we consider the case I mentioned (supporting both numbers and tensors / lists in torch functions), then we would want `torch.as_tensor` to not break the computation graph. Thus, `torch.as_tensor(tensor, dtype=dtype, device=device)` should work as `.to(dtype, device)`, for tensors.

Comment (apaszke) [reply torch/csrc/utils/tensor_new.cpp]: Yeah I think the primary use-case for `as_tensor` is writing out functions that can work both with tensors and with regular Python objects, and there's no reason to make them non-differentiable if the inputs already were tensors. I really think it should preserve the graph, and use `.to` to casting if necessary.

Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: @apaszke @gchanan @fmassa Thanks a lot for the very helpful insights! I will keep the differentiable behavior of `torch.as_tensor()` as documented.

File: torch/csrc/utils/tensor_new.cpp (hunk 5 above)
Comment (weiyangfb) [reply torch/csrc/utils/tensor_new.cpp]: Can I keep the warning at the caller function of `internal_new_from_data`? Otherwise I will need to put a default name at `internal_new_from_data` that might be accidentally picked up when a caller function triggers the warning..

**Comment by weiyangfb** [comment]
@SsnL Yes, it warns because `torch.sparse_coo_tensor` calls into `internal_new_from_data`, where I put the warning previous. I removed it at this PR, so it should be gone after this lands.

**Review by apaszke — APPROVED** [review]
